x-pack/plugins/maps/public/store/map.js

```diff
@@ -359,3 +363,43 @@ function getValidDataRequest(state, action, checkRequestToken = true) {
 function findLayerById(state, id) {
   return state.layerList.find(layer => layer.id === id);
 }
+
+function rollbackAnyTrailingTrackedLayerState(state) {
+  state.layerList.forEach(layer => {
+    if (layer[TRACKED_LAYER_DESCRIPTOR]) {
+      rollbackTrackedLayerState(state, layer.id);
+    }
+  });
+}
+
+function trackCurrentLayerState(state, layerId) {
+  const layer = findLayerById(state, layerId);
+  const layerCopy = copyPersistentState(layer);
+  return updateLayerInList(state, layerId, TRACKED_LAYER_DESCRIPTOR, layerCopy);
+}
+
+function removeTrackedLayerState(state, layerId) {
+  const layer = findLayerById(state,  layerId);
```

Comment:
state should never be mutated. That could lead to very hard to figure out bugs. This method needs to return the modified state without effecting the original